HCL generation problems

The HCL generation still has some errors and sometimes fails or produces wrong outputs (not right directions or nodes).

We do not have such issue on TFState (we use to but was fixed) but on HCL we have it.

We have to find why is it failing, most probably related on how the HCL->`graph.Graph` before any operation is done to the graph as reading the HCL it's not as easy as TFState(json).

This issues are "hard" to reproduce as they fail once in a while but not always, for example:

```
--- FAIL: TestFromHCL_FlexibleEngine (0.00s)
    --- FAIL: TestFromHCL_FlexibleEngine/SuccessSG (0.00s)
        helper_test.go:66: 
            	Error Trace:	helper_test.go:66
            	            				hcl_test.go:90
            	Error:      	Fail
            	Test:       	TestFromHCL_FlexibleEngine/SuccessSG
            	Messages:   	The Edge with Source "flexibleengine_compute_instance_v2.instance_two" and Target "flexibleengine_compute_instance_v2.instance_one" is present but in the other direction
FAIL
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,12 @@
   ([Issue #13](https://github.com/cycloidio/inframap/issues/13))
 - Google graph generation from TFState
   ([Issue #7](https://github.com/cycloidio/inframap/issues/7))
+- Google graph generation from HCL
+  ([Issue #27](https://github.com/cycloidio/inframap/issues/27))
+### Fixed
+
+- HCL generation errors
+  ([Issue #29](https://github.com/cycloidio/inframap/issues/29))
 
 ## [0.1.1] _2020-07-16_
 

--- a/generate/hcl.go
+++ b/generate/hcl.go
@@ -11,6 +11,7 @@ import (
 	"github.com/hashicorp/hcl/v2/hclsyntax"
 	"github.com/hashicorp/hcl/v2/hclwrite"
 	"github.com/hashicorp/terraform/configs"
+	"github.com/hashicorp/terraform/configs/hcl2shim"
 	uuid "github.com/satori/go.uuid"
 	"github.com/spf13/afero"
 )
@@ -148,6 +149,12 @@ func FromHCL(fs afero.Fs, path string, opt Options) (*graph.Graph, error) {
 		}
 	}
 
+	// call the preprocess method for each
+	// TF provider in the file
+	if err := preprocess(g, resourcesRawConfig, opt); err != nil {
+		return nil, err
+	}
+
 	if opt.Clean {
 		g.Clean()
 	}
@@ -201,22 +208,30 @@ func getBodyJSON(b *hclsyntax.Body) map[string]interface{} {
 		v, _ := attrv.Expr.Value(nil)
 		t := v.Type().FriendlyName()
 		switch t {
-		case "string", "bool", "number":
+		case "string", "bool", "number", "dynamic":
 			for i, vr := range attrv.Expr.Variables() {
 				if i > 0 {
 					continue
 				}
 				links[attrk] = fmt.Sprintf("${%s}", string(hclwrite.TokensForTraversal(vr).Bytes()))
 			}
+			if _, ok := links[attrk]; !ok {
+				links[attrk] = hcl2shim.ConfigValueFromHCL2(v)
+			}
 		case "tuple":
 			aux := make([]interface{}, 0)
 			for _, vr := range attrv.Expr.Variables() {
 				aux = append(aux, fmt.Sprintf("${%s}", string(hclwrite.TokensForTraversal(vr).Bytes())))
 			}
-			// We continue to not add empty information to the config
-			// so it's clean and only has required information
 			if len(aux) == 0 {
-				continue
+				i := hcl2shim.ConfigValueFromHCL2(v)
+				islice, ok := i.([]interface{})
+				if !ok {
+					// We continue to not add empty information to the config
+					// so it's clean and only has required information
+					continue
+				}
+				aux = islice
 			}
 			links[attrk] = aux
 		}

--- a/generate/hcl_test.go
+++ b/generate/hcl_test.go
@@ -91,6 +91,38 @@ func TestFromHCL_FlexibleEngine(t *testing.T) {
 	})
 }
 
+func TestFromHCL_Google(t *testing.T) {
+	t.Run("Success", func(t *testing.T) {
+		fs := afero.NewOsFs()
+
+		g, err := generate.FromHCL(fs, "./testdata/google_hcl.tf", generate.Options{Clean: true, Connections: true})
+		require.NoError(t, err)
+		require.NotNil(t, g)
+
+		eg := &graph.Graph{
+			Nodes: []*graph.Node{
+				&graph.Node{
+					Canonical: "google_compute_instance.inframap-tmp-two",
+				},
+				&graph.Node{
+					Canonical: "google_compute_instance.inframap-tmp",
+				},
+			},
+			Edges: []*graph.Edge{
+				&graph.Edge{
+					Target: "google_compute_instance.inframap-tmp",
+					Source: "google_compute_instance.inframap-tmp-two",
+					Canonicals: []string{
+						"google_compute_firewall.allow-ssh",
+					},
+				},
+			},
+		}
+
+		assertEqualGraph(t, eg, g, nil)
+	})
+}
+
 func TestFromHCL_Module(t *testing.T) {
 	t.Run("SuccessSG", func(t *testing.T) {
 		fs := afero.NewOsFs()

--- a/generate/state.go
+++ b/generate/state.go
@@ -569,6 +569,10 @@ func preprocess(g *graph.Graph, cfg map[string]map[string]interface{}, opt Optio
 				Target: edge[1],
 			})
 			if err != nil {
+				// If the edge already exists we can ignore it
+				if errors.Is(err, errcode.ErrGraphAlreadyExistsEdge) {
+					continue
+				}
 				return fmt.Errorf("could not add edge: %w", err)
 			}
 		}

--- a/generate/testdata/google_hcl.tf
+++ b/generate/testdata/google_hcl.tf
@@ -1,0 +1,88 @@
+resource "google_compute_instance" "inframap-tmp" {
+  name         = "inframap-tmp"
+  machine_type = "f1-micro"
+  zone         = "us-east1-b"
+
+  tags = ["front", "ssh"]
+
+  boot_disk {
+    initialize_params {
+      image = "debian-cloud/debian-9"
+    }
+  }
+
+  scheduling {
+    preemptible       = true
+    automatic_restart = false
+  }
+
+  network_interface {
+    network = google_compute_network.vpc_network.name
+
+    access_config {
+      // Ephemeral IP
+    }
+  }
+
+}
+
+resource "google_compute_instance" "inframap-tmp-two" {
+  name         = "inframap-tmp-two"
+  machine_type = "f1-micro"
+  zone         = "us-east1-b"
+
+  tags = ["ssh"]
+
+  boot_disk {
+    initialize_params {
+      image = "debian-cloud/debian-9"
+    }
+  }
+
+  scheduling {
+    preemptible       = true
+    automatic_restart = false
+  }
+
+  network_interface {
+    network = google_compute_network.vpc_network.name
+
+    access_config {
+      // Ephemeral IP
+    }
+  }
+
+}
+
+resource "google_compute_network" "vpc_network" {
+  name                    = "inframap-network-tmp"
+  auto_create_subnetworks = true
+}
+
+resource "google_compute_firewall" "allow-http" {
+  network     = google_compute_network.vpc_network.name
+  description = "allow incoming http traffic"
+  name        = "allow-http"
+
+  allow {
+    protocol = "tcp"
+    ports    = ["80"]
+  }
+  direction     = "INGRESS"
+  source_ranges = ["0.0.0.0/0"]
+  target_tags   = ["front"]
+}
+
+resource "google_compute_firewall" "allow-ssh" {
+  network     = google_compute_network.vpc_network.name
+  description = "allow SSH between instances"
+  name        = "allow-ssh"
+
+  allow {
+    protocol = "tcp"
+    ports    = ["22"]
+  }
+  direction   = "INGRESS"
+  source_tags = ["ssh"]
+  target_tags = ["front"]
+}

